cobald_hep_plugins/timer.py

```diff
@@ -0,0 +1,92 @@
+"""Time-based demand control decorators and helpers."""
+
+from __future__ import annotations
+
+import asyncio
+from datetime import datetime, time as datetime_time
+from typing import Mapping, Union, Iterable
+
+from cobald.daemon import service
+from cobald.interfaces import Pool, PoolDecorator
+
+TimeInput = Union[str, datetime_time]
+Schedule = Mapping[TimeInput, float]
+
+
+def str_to_time(value: str) -> datetime_time:
+    """Convert a HH:MM string into a :class:`datetime.time` object."""
+    try:
+        hours_str, minutes_str = value.strip().split(":", 1)
+        hours, minutes = int(hours_str), int(minutes_str)
+    except ValueError as exc:
+        raise ValueError(f"invalid time specification {value}")
+    return datetime_time(hour=hours, minute=minutes)
+
+
+def latest_timestamp(
+    times: Iterable[datetime_time],
+    *,
+    reference: datetime_time | None = None, # easier to test with
+) -> datetime_time:
+    """Return the latest timestamp with respect to the current time."""
+    ordered_times = sorted(times) # ascending
+    current_time = datetime.now().time() if reference is None else reference
+    for candidate in reversed(ordered_times): # descending
+        if candidate <= current_time:
+            return candidate
+    # All configured times are in the future relative to ``current_time``.
+    # return latest timestamp
+    return ordered_times[-1]
```

Comment:
This is probably no longer needed, see later comments. If you want to keep it:
- `current_timestamp` or `recent_timestamp` is probably a better name (without knowing the implementation, "latest" is likely to be read as "most late" instead of "most recent").
-  ordering should already be a prerequisite and established outside this function. Sorting the input again and again is a waste since a) the actual user input is likely sorted anyways and b) it's simple to sort once when reading the input.